### ISSUE TITLE
CLI/Lib Refactor

### ISSUE THREAD

**Issue by dustinlacewell** [TypeScriptToLua/TypeScriptToLua#393]
## Motive
Currently there is not enough seperation between the implementation of parts within TSTL. While TSTL does offer a library-like interface this is mostly incidental. TSTL was originally designed with command-line usage in mind.

## Objective
Let's identify and discuss changes that will strengthen a separation of concerns between the core implementation, and those bits which implement the CLI

- *Core*: Everything to do with the business of transpilation
- *CLI*: Everything to do with translating command-line arguments and execution environment (variables) into parametric calls into core and reporting issues back to users through stderr/logging, etc.

## Core Should Be Easily Composed

Here is a pseudo-code example of what may be possible. The methods of `Parser`, `Transformer` and `FileEmitter` and variants all return some structured object which is suitable for consumption by the next applicable processor in the pipeline making them easy to compose together as shown. Each `...` in the example is a point where configuration can be injected, including plugin objects. Additionally, these objects can be subclassed for more control. Finally, they can be replaced entirely for ultimate control. 

Missing here is exception handling clauses which would catch exception variants raised from inside core. Each client application is more easily able to handle all aspects of using TSTL this way. The CLI implementation should also become more focused and narrow as a result.

```typescript
import * as tstl from 'tstl';

const
    myInputFiles: File[],
    parser = tstl.Parser(...),
    transformer = tstl.Transformer(...),
    emitter = tstl.FileEmitter(...);

for (const file of myInputFiles) {
    for (const ast of parser.parseFile(f)) {
        const result = transformer.transform(ast);
        // emitter(result);
        customEmissionHandler(result, ...);
    });
}
```

### Core Catches Exceptions

Currently there are core-ish routines which catch and handle exceptions on behalf of the CLI use-case. This is problematic for library use-cases wherein those errors need to be handled specifically by calling code.

**Cases**:
https://github.com/Perryvw/TypescriptToLua/blob/master/src/LuaTranspiler.ts#L115

### Core Interacts With Process

Core routines should not be in the business of doing things like calling `process.exit` or participating in any other process management such as watchers.

**Cases**:
https://github.com/Perryvw/TypescriptToLua/blob/master/src/LuaTranspiler.ts#L49

### Reporting

The core should never directly utilize functions like `console.log` to report information to the user. Instead the core should return structured information to the caller. TSTL's CLI can interpret and format this information for outputting to the console or any other log destination. Similiarly, library users can handle reporting in the way that makes sense in the context of each use-case (such as IDE extensions).

**Cases**:
https://github.com/Perryvw/TypescriptToLua/blob/master/src/LuaTranspiler.ts#L147
https://github.com/Perryvw/TypescriptToLua/blob/master/src/LuaTranspiler.ts#L149

### Core Parametricity

Core routines should take in all of their configuration through parameter arguments and should never reach out to the environment or depend on configuration files. The CLI should be responsible for translating configuration artifacts, environment variables and commnand-line arguments into the correct core routine arguments.

**Comment by lolleko** [comment]
duplicate of #371 closing #371 because this one is more detailed

**Comment by Perryvw** [comment]
Totally agree on the first two points, especially since this seems to be done purely for CLI reasons. That should be a relatively small change. I am currently having my doubts if reportErrors belongs in the transpiler at all, will think about that a little.

The last point about core parameters seems pretty vague, it would be nice to make this a bit more concrete. 
One thing I did run into was LuaTranspiler always grabbing options from the program parameter at: https://github.com/Perryvw/TypescriptToLua/blob/master/src/LuaTranspiler.ts#L22

A better solution is probably to add an optional `options` argument to the constructor, and falling back on `program.getOptions()` when the constructor argument is omitted.

**Comment by dustinlacewell** [comment]
>> The last point about core parameters seems pretty vague

For sure. I remember when working with the code there was at some point some option that was being taken from somewhere I couldn't directly affect as the caller into that code. But I can't remember exactly what it was... If I end up not being able to remember I'll go ahead and remove this item. Though, it is actually a good principle to "have the board" so to speak as it it is in the spirit of the overall issue.

**Comment by dustinlacewell** [comment]
I updated the issue description with a new section **Core Should Be Easily Composed** to demonstrate the overal aesthetic(?) or intention behind the changes. There is some example code which shows what the basic use of TSTL would look like from a library use-case perspective. But this is also what the CLI implementation could also look like. I think it hints at a very easy to write beginner's tutorial as well.

**Comment by dustinlacewell** [comment]
> One thing I did run into was LuaTranspiler always grabbing options from the program parameter

@Perryvw Assume we grow a module which is focused just on dealing with TS. That is, instantiating a `ts.Program` and then providing the resulting `SourceFile`s back to the caller of our module. Is there an argument that we should hide TS completely within the implementation of that module and not expose it as part of the API of the TSTL module itself? I think the complication with the seperation of concerns here is the `-p` flag and how TS can get some of it's options from `tsconfig.json` and if it does, then TSTL doesn't know about them -- that's why TSTL has to get the options from the `ts.Program`. As long as we allow TS to utilize `tsconfig.json` vs loading `tsconfig.json` ourselves and passing options to TS manually, we'll always have this sort of shared authority over what the actual options in play are. I'm not sure it's worth the effort to do that unless `tsconfig.json` options are named the same as their CLI flags or something in which case it would actually be trivial to go that route. Thoughts?

**Comment by Perryvw** [comment]
It depends on what we settle on for the rest of the API. I think it is strange we pack options into `ts.Program` and then immediately unpack it again. I would like to keep open the possibility of calling the same transpilation process with different options, or overriding the default options.

Arguably this is already possible with `compileWithOptions`, but the way in which seems a little awkward to me. Like I said this also kind of relies on what we do with the rest of the API, so I think we revisit this once that is clearer.

**Comment by Perryvw** [comment]
As a recap of yesterday's discussion on Discord, the current setup for building a tstl pipeline in user code would be:
```ts
const fileNames = [];
const options = {};

const program = ts.createProgram(fileNames, options);
const transformer = new LuaTransformer(program, options);
const printer = new LuaPrinter(options);

for (const sourceFile of program.getSourceFiles())
{
    const [luaAST, luaLibSettings] = transformer.transformSourceFile(sourceFile);

    const lua = printer.print(luaAST, luaLibSettings);
    writeFileSync(sourceFile.fileName.replace(".ts", ".lua"), lua);
}
```
This looks fine to me, except that we might be missing a way to load a project and tsconfig. So we should make sure every component here is exported, as well as all the components required to implement an extension of LuaTransformer or LuaPrinter.

Then the one thing I am not entirely sure about is how and from where we are going to expose the project or tsconfig reading.

**Comment by dustinlacewell** [comment]
In my experimental fork, I have created new module `Parser.ts` which hides `ts` module completely from the user. You pass the filename of the `tsconfig.json` and it uses the `ts` compiler routines to load the `tsconfig.json` into a `CompilerOptions` and uses that to create the `Program`. 

This way tstl is the configuration authority and ts is just an implementation detail users don't need to know anything about. I know we talked about it on Discord, but I prefer it much more.
